tests/test_primitives/test_memory.py

```diff
@@ -19,10 +19,8 @@ class test_memory_basic(m.Circuit):
 
         )
         Mem4x5 = m.Memory(4, m.Bits[5])()
-        Mem4x5.RADDR @= io.raddr
-        io.rdata @= Mem4x5.RDATA
-        Mem4x5.WADDR @= io.waddr
-        Mem4x5.WDATA @= io.wdata
+        io.rdata @= Mem4x5[io.raddr]
+        Mem4x5[io.waddr] = io.wdata
```

Comment:
One thought, here we use `=` since it dispatched to __setitem__, alternatively we could look into `__getitem__` returning something smart so that we could `Mem4x5[io.waddr] @= io.wdata` as the syntax, but this is a bit tricky since we'll need to distinguish between cases where we're doing reads and writes.  The benefit would be we would not need to introduce `=` in this special context (I think ideally we could just state that `@=` is alway used for working with magma values)

Reply:
That sounds nice. One idea would be to return a special object which stages the decision of read or write. This special object could look like a normal value when something wires to it (and then do the read op.). If it is wired to, then it would be treated as read (and we could ensure it isn't written to again). Probably some hiccups to what I just proposed but just a thought.